### DISCUSSION TITLE
Configure zwave logs to output in json format?

### DISCUSSION THREAD

**Discussion by zestysoft** [zwave-js/zwave-js#5214] [Q&A] [answered]
I'm having trouble figuring out how to parse these debug logs in their current "human" readable format with indentations, etc -- this would be much easier to filter through if the data was in json format, or at least in a format where all the data is contained within a single line.  I was hoping there might be a configuration switch somewhere to enable json output, or at least some kind of greppable format?

**Comment by kpine** [comment, marked as answer]
Customizing the logging output is done by installing [log transports](https://github.com/zwave-js/log-transports). There are some default ones, including [JSON](https://github.com/zwave-js/log-transports/tree/master/packages/json), bundled with Z-Wave JS, or you can create your own.

I think usage of these log transports is intended for application developers, not end users. You probably want to make a feature request to the application you're using (e.g. Z-Wave JS UI) to support this functionality, unless you're a developer yourself. The log transport configuration is part of the driver options, which most applications don't give you direct access to. Popular applications like Z-Wave JS UI or the Home Assistant add-on aren't currently capable of configuring custom log transports.

I experimented with the bundled JSON transport at one point, and had a configuration module that looked something like this:

```javascript
const logjson = require('@zwave-js/log-transport-json')

function getLogConfig() {
  jsonlog = new logjson.JSONTransport()
  jsonlog.stream.on('data', console.log)

  return { transports: [ jsonlog ] };
}

module.exports = {
  logConfig: getLogConfig(),
};
```

It outputs the JSON formatted logs to the console. There is probably a better way to do this (I'm no JS expert), and this is limited to console output only.

The node-red-contrib-zwave-js project provides a real example of using custom log transports: https://github.com/zwave-js/node-red-contrib-zwave-js/blob/e9bf1697c60ae5d0cc8b7098b779947056674164/zwave-js/zwave-js.js#L238

**Reply by zestysoft** [reply]
Hmm -- I've got the container zwavejs/zwavejs2mqtt:8.4.1 running in a pod in kubernetes rn and am happy to exec into the container and alter files in there to eek out the json output -- can you describe a bit more about where I'd add that configuration module you've written?  That looks like exactly what I need -- I'm perfectly happy working from the terminal.

**Reply by kpine** [reply]
zwave-js UI doesn't support that. You'd have to modify the application code.

**Reply by zestysoft** [reply]
Right -- that's what I was suggesting above -- I have direct access to the application code within the running container.  I guess it's not as simple as editing a js file somewhere to add the code you've shared above.

**Reply by kpine** [reply]
zwave-js-ui is written in Typescript, it gets compiled down to Javascript. Sorry, but I wouldn't be able to explain how to modify the code. You'll have to search around the compiled code for the driver option code I referenced above and make the changes.

**Reply by zestysoft** [reply]
Sorry -- maybe I'm in the wrong repo then -- node-zwave-js == zwave-js-ui?  I'm confused why you keep brining up the UI.  The container has a UI component, but not only that.  Which repo is for the non-ui stuff?

**Reply by kpine** [reply]
Z-Wave JS UI is the software you're using, formerly zwavejs2mqtt. https://github.com/zwave-js/zwave-js-ui

Z-Wave JS (node-zwave-js) is the driver software.

The application software is responsible for configuring the logging, aside from the defaults. That would include other logging transports.

**Reply by kpine** [reply]
In fact, there is already an issue related to this. https://github.com/zwave-js/zwave-js-ui/issues/2136

**Reply by zestysoft** [reply]
Thanks for clearing this up for me. I'll mark that as the answer in case someone else finds this thread and wants to follow progress.

**Comment by kpine** [comment]
I think it would be a nice enhancement if the driver configuration would allow switching between any of the official extra log formats, e.g. JSON or logfmt output, which a simple option, instead of requiring code to enable the transports, while supporting the existing file/console output and log rotation. Then all applications could take advantage of the different formats.
